### PR TITLE
fix(freespace_planning_algorithms): throw error when no waypoints stored

### PR DESCRIPTION
**Pull request by HiroIshida**
### 1. safer access to waypoints
Current implementation of `compute_length` is problematic because it can invalidly access vector element and cause the following error. 
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 0) >= this->size() 
```
This is mainly happens when solver failed to solve, which means no waypoints are set.  So, also to fix the root cause, I add check if the waypoints is empty in `getWaypoints` function.

### 2. update test
related to 1, I updated the test so that if planning fails, don't call `getWaypoints`

### note
This may related to https://github.com/autowarefoundation/autoware.universe/issues/2439 but if invalid access in `compute_length()` causes the error, that's mean it takes more than 10 seconds to solve a single problem by astar, and the chance is quite low. So, probably this PR will not fix https://github.com/autowarefoundation/autoware.universe/issues/2439 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp [hunk 1]
```diff
@@ -110,7 +110,14 @@ class AbstractPlanningAlgorithm
   virtual bool makePlan(
     const geometry_msgs::msg::Pose & start_pose, const geometry_msgs::msg::Pose & goal_pose) = 0;
   virtual bool hasObstacleOnTrajectory(const geometry_msgs::msg::PoseArray & trajectory) const;
-  const PlannerWaypoints & getWaypoints() const { return waypoints_; }
+  const PlannerWaypoints & getWaypoints() const
+  {
+    if (waypoints_.waypoints.size()) {
```
Comment (TakaHoribe) [inline planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp:115]:
```suggestion
    if (!waypoints_.waypoints.empty()) {
```

File: planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp [hunk 2]
```diff
@@ -110,7 +110,14 @@ class AbstractPlanningAlgorithm
...
+      return waypoints_;
+    }
+    throw std::runtime_error("waypoints has length 0. Did you check that problem solved?");
```
Comment (TakaHoribe) [inline planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp:118]: It is natural to return an empty waypoint even if it is empty. The error handling should be done outside of this function

**Review by TakaHoribe — COMMENTED** [review]
@HiroIshida Thank you for this PR. 
Question: What will happen when the solver fails to compute the waypoint? Before it results in an invalid index access exception, and with this PR the planner still dies with another exception?

File: planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp (hunk 1 above)
Comment (HiroIshida) [reply planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp:115]: Sorry this was a typo

File: planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp (hunk 2 above)
Comment (HiroIshida) [reply planning/freespace_planning_algorithms/include/freespace_planning_algorithms/abstract_algorithm.hpp:118]: OK, I should remove this

**Comment by HiroIshida** [comment]
@TakaHoribe 

> Before it results in an invalid index access exception, and with this PR the planner still dies with another exception?
Yes. In the previous implementation, when user and developer mistakenly call `compute_length` against empty waypoint, it the program finish with the following error. 
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 0) >= this->size() 
```
And although the error is simple, typically user need to compile it with debug flag and have to use debugger to detect this error. But if we throw a error with explicit message, user can easily detect the reason of bug without any debugger.

**Comment by TakaHoribe** [comment]
@HiroIshida 

> mistakenly call compute_length against empty waypoint

Does this operation happen accidentally? For example, in the case of path search failure?

**Review by TakaHoribe — APPROVED** [review]
LGTM

**Pull request merged**